File: bindings/nim/src/proven/safe_path.nim
# SPDX-License-Identifier: PMPL-1.0-or-later
# Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <jonathan.jewell@open.ac.uk>
#
# Safe filesystem path operations with traversal attack prevention.
# Thin wrapper over libproven FFI -- all logic lives in Idris.

import std/options
import lib_proven

proc hasTraversal*(path: string): bool =
  ## Check if a path contains directory traversal sequences.
  if path.len == 0:
    return false
  let res = provenPathHasTraversal(unsafeAddr path[0], csize_t(path.len))
  if res.status == PROVEN_OK:
    return res.value
  false

proc isSafe*(path: string): bool =
  ## Check if a path is safe (no traversal attacks).
  not hasTraversal(path)

proc sanitizeFilename*(filename: string): Option[string] =
  ## Sanitize a filename by removing dangerous characters.
  if filename.len == 0:
    return some("")
  let res = provenPathSanitizeFilename(unsafeAddr filename[0], csize_t(filename.len))
  if res.status == PROVEN_OK and res.value != nil:
    let sanitized = $res.value
    provenFreeString(res.value)
    return some(sanitized)
  none(string)

proc safeJoin*(base: string, parts: varargs[string]): Option[string] =
  ## Safely join path components, rejecting traversal attempts.
  ## Delegates traversal checking to libproven, then joins paths.
  for part in parts:
    if hasTraversal(part):
      return none(string)

  var joined = base
  for part in parts:
    let sanitized = sanitizeFilename(part)
    if sanitized.isNone:
      return none(string)
    if joined.len > 0 and joined[^1] != '/':
      joined.add('/')
    joined.add(sanitized.get)

  some(joined)

File: bindings/lua/proven/safe_path.lua
-- SPDX-License-Identifier: Apache-2.0
-- SPDX-FileCopyrightText: 2025 Hyperpolymath

--- Safe filesystem path operations with traversal attack prevention.
--- @module proven.safe_path

local safe_path = {}

--- Check if a path contains directory traversal sequences.
--- @param path string
--- @return boolean
function safe_path.has_traversal(path)
    return path:find("%.%.") ~= nil or path:find("~") ~= nil
end

--- Check if a path is safe (no traversal attacks).
--- @param path string
--- @return boolean
function safe_path.is_safe(path)
    return not safe_path.has_traversal(path)
end

--- Sanitize a filename by removing dangerous characters.
--- @param filename string
--- @return string
function safe_path.sanitize_filename(filename)
    return filename
        :gsub("%.%.", "_")
        :gsub("/", "_")
        :gsub("\\", "_")
        :gsub("<", "_")
        :gsub(">", "_")
        :gsub(":", "_")
        :gsub('"', "_")
        :gsub("|", "_")
        :gsub("%?", "_")
        :gsub("%*", "_")
        :gsub("%z", "_")
end

--- Safely join path components, rejecting traversal attempts.
--- @param base string
--- @param parts table Array of path components
--- @return string|nil Returns nil if traversal detected
function safe_path.safe_join(base, parts)
    for _, part in ipairs(parts) do
        if safe_path.has_traversal(part) then
            return nil
        end
    end

    local sanitized = {}
    for _, part in ipairs(parts) do
        table.insert(sanitized, safe_path.sanitize_filename(part))
    end

    local path = base
    for _, part in ipairs(sanitized) do
        if path:sub(-1) == "/" then
            path = path .. part
        else
            path = path .. "/" .. part
        end
    end

    return path
end

return safe_path
